k8s/secrets.go

```diff
@@ -0,0 +1,261 @@
+	// GetSecrets queries Kubernetes for a list of secrets by name in the given k8s namespace.
+	// This should only be used if you need access to the actual secret structure/value. Specifically,
+	// inside the FunctionFactory.
+	GetSecrets(ctx context.Context, namespace string, secretNames []string) (map[string]*apiv1.Secret, error)
+}
+
+// SecretsInterfacer exposes the SecretInterface getter for the k8s client.
+// This is implemented by the CoreV1Interface() interface in the Kubernetes client.
+// The SecretsClient only needs this one interface, but needs to be able to set the
+// namespaces when the interface is instantiated, meaning, we need the Getter and not the
+// SecretInterface itself.
+type SecretInterfacer interface {
+	// Secrets returns a SecretInterface scoped to the specified namespace
+	Secrets(namespace string) typedV1.SecretInterface
+}
+
+type secretClient struct {
+	kube SecretInterfacer
+}
+
+// NewSecretsClient constructs a new SecretsClient using the provided Kubernetes client.
+func NewSecretsClient(kube kubernetes.Interface) SecretsClient {
+	return &secretClient{
+		kube: kube.CoreV1(),
+	}
+}
+
+func (c secretClient) List(ctx context.Context, namespace string) (names []string, err error) {
+	res, err := c.kube.Secrets(namespace).List(c.selector())
+	if err != nil {
+		log.Printf("failed to list secrets in %s: %v\n", namespace, err)
+		return nil, err
+	}
+
+	names = make([]string, len(res.Items))
+	for idx, item := range res.Items {
+		// this is safe because size of names matches res.Items exactly
+		names[idx] = item.Name
+	}
+	return names, nil
+}
+
+func (c secretClient) Create(ctx context.Context, secret types.Secret) error {
+	err := c.validateSecret(secret)
+	if err != nil {
+		return err
+	}
+
+	req := &apiv1.Secret{
+		Type: apiv1.SecretTypeOpaque,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secret.Name,
+			Namespace: secret.Namespace,
+			Labels: map[string]string{
+				secretLabel: secretLabelValue,
+			},
+		},
+		StringData: map[string]string{
+			secret.Name: secret.Value,
+		},
+	}
+
+	_, err = c.kube.Secrets(secret.Namespace).Create(req)
+	if err != nil {
+		log.Printf("failed to create secret %s.%s: %v\n", secret.Name, secret.Namespace, err)
+		return err
+	}
+
+	log.Printf("created secret %s.%s\n", secret.Name, secret.Namespace)
+
+	return nil
+}
+
+func (c secretClient) Replace(ctx context.Context, secret types.Secret) error {
+	err := c.validateSecret(secret)
+	if err != nil {
+		return err
+	}
+
+	kube := c.kube.Secrets(secret.Namespace)
+	found, err := kube.Get(secret.Name, metav1.GetOptions{})
+	if err != nil {
+		log.Printf("can not retrieve secret for update %s.%s: %v\n", secret.Name, secret.Namespace, err)
+		return err
+	}
+
+	found.StringData = map[string]string{
+		secret.Name: secret.Value,
+	}
+	_, err = kube.Update(found)
+	if err != nil {
+		log.Printf("can not update %s.%s: %v\n", secret.Name, secret.Namespace, err)
```

Comment:
Perhaps it'd be more informative in the logs as: `can not update secret: ` instead of `can not update`?